### PR TITLE
Add catchAnalyzer precommit task

### PR DESCRIPTION
**Pull request by rmuir**
The idea here is to add a simple check to find broken catch blocks. It detects a couple of common problems:
- throwing a different exception, but forgetting to pass the original one to its ctor, or calling initCause/addSuppressed/etc. So information about what originally happened is lost!
- swallowing an exception completely (e.g. just logging it and moving on).

It does some flow analysis (+ hacks) to track the original exception via all possible codepaths. If the method can return without throwing anything, or throws a different exception that doesn't have the original chained to it in some way, it fails.

It also knows about rethrowers that we have, and some helper methods "boxers" that wrap exceptions in other ones.

Scans are true to the bytecode: so don't be surprised by the errors for broken catch blocks with try-with-resources/finally/etc. These aren't duplicates, that's what javac is actually doing with your code.

I think if we really want to swallow an exception, its ok to annotate the method with `@SwallowsException` saying that its intentional. Otherwise in many cases its a bug.

Unfortunately, there are thousands of violations. Even just annotating them all with `@SwallowsException(reason = "?")` takes forever (I did some of them, but quickly gave up). We have to figure out a way to iteratively clean this up...

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java [hunk 1]
```diff
@@ -204,7 +205,7 @@ public void writeTo(StreamOutput out) throws IOException {
     }
 
 
-    @Override
+    @Override @SwallowsExceptions(reason = "?")
     public String toString() {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java]: One mindless way to solve the problem with methods that look like this is to forward them all to `Strings.toString` and just annotate that one with `@SwallowsException(reason = "We don't want a broken toString to bubble up")`. I'm not sure that is a **good** reason, but it is the reason the `toString`s all do this.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java [hunk 2]
```diff
@@ -203,7 +204,7 @@ void getFinishedTaskFromIndex(Task thisTask, GetTaskRequest request, ActionListe
                 request.getTaskId().toString());
         get.setParentTask(clusterService.localNode().getId(), thisTask.getId());
         client.get(get, new ActionListener<GetResponse>() {
-            @Override
+            @Override @SwallowsExceptions(reason = "?")
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java]: These are expected, even required when implementing `ActionListener`. Either we should create an `AbstractActionListener` that does this for us the same way we have a `AbstractRunnable` or we should add a permanent hack to allow this specific construct in `ActionListener` subclasses. The former seems like a better choice but it'd mean more work before we can get this merged.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java [hunk 3]
```diff
@@ -74,7 +75,7 @@ protected ClusterBlockException checkBlock(GetSnapshotsRequest request, ClusterS
         return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
     }
 
-    @Override
+    @Override @SwallowsExceptions(reason = "?")
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: This is another place where we should probably have a common superclass that does the forwarding to the listener.

File: core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java [hunk 4]
```diff
@@ -372,7 +373,7 @@ public CreateIndexRequest source(BytesReference source) {
     /**
      * Sets the settings and mappings as a single source.
      */
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings("unchecked") @SwallowsExceptions(reason = "?")
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java]: This one should probably just include the exception as the cause.

File: settings.gradle [hunk 5]
```diff
@@ -18,6 +18,7 @@ List projects = [
   'test:fixtures:example-fixture',
   'test:fixtures:hdfs-fixture',
   'test:logger-usage',
```
Comment (nik9000) [inline settings.gradle]: If you are going to make a `:test:tools` then we should move logger-usage into it eventually. Probably in another PR.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java (hunk 2 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java]: I dont think we should try to fix any of the problems here? it would be great to simply document them all with the annotation and try to fix them separately. there are still 700 un-annotated violations in es core alone. I tried to start annotating them, its too exhausting.

File: test/tools/catch-analyzer/build.gradle [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+dependencies {
+  compile 'org.ow2.asm:asm-debug-all:6.0_ALPHA' // use asm-debug-all because WE LIKE LINE NUMBERS!
+  testCompile "com.carrotsearch.randomizedtesting:randomizedtesting-runner:${versions.randomizedrunner}"
+  testCompile "junit:junit:${versions.junit}"
+  testCompile "org.hamcrest:hamcrest-all:${versions.hamcrest}" // hamcrest is evil, but randomized runner no worky without
+}
+
+// some things are disabled as they only work if you have es/lucene
+loggerUsageCheck.enabled = false
+namingConventions.enabled = false
```
Comment (nik9000) [inline test/tools/catch-analyzer/build.gradle]: namingConventions defaults to checking lucene classes but you can configure it to use any base class you like. The client project does it.

File: test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/CatchAnalyzer.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            if ((reader.getAccess() & Opcodes.ACC_SYNTHETIC) != 0) {
+                continue;
+            }
+            ClassAnalyzer analyzer = new ClassAnalyzer(reader.getClassName(), loader);
+            reader.accept(analyzer, 0);
+            scannedCount++;
+            
+            violationCount += analyzer.logViolations(System.out);
+        }
+        long endTime = System.currentTimeMillis();
```
Comment (nik9000) [inline test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/CatchAnalyzer.java]: It is probably better to use nanoTime here so you don't get clock skew giving you weird numbers. Not like it matters a whole lot here though.

File: test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/ClassAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                stream.println(sb.toString());
+                violationCount++;
+            }
+        } else {
+            for (Map.Entry<Method,MethodAnalyzer> entry : analyses.entrySet()) {
+                Method method = entry.getKey();
+                MethodAnalyzer analysis = entry.getValue();
+                if (analysis.suppressed) {
+                    if (analysis.violations.isEmpty()) {
+                        StringBuilder sb = createMessage("Wrong method annotation", method, -1);
```
Comment (nik9000) [inline test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/ClassAnalyzer.java]: Should this string be "Method marked as `@SwallowsException` but doesn't swallow any exceptions" or am reading this wrong?

File: test/tools/catch-analyzer/src/test/java/org/elasticsearch/test/BaseTestCase.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.util.concurrent.atomic.AtomicLong;
+
+@TestMethodProviders({
+    LuceneJUnit3MethodProvider.class,
+    JUnit4MethodProvider.class
+})
+@RunWith(RandomizedRunner.class)
+public class BaseTestCase extends Assert {
```
Comment (nik9000) [inline test/tools/catch-analyzer/src/test/java/org/elasticsearch/test/BaseTestCase.java]: I don't know that we should fix this now, but I think failures of this test will miss the gradle reproduction steps, right? I've been thinking of pulling those into a tiny shared project without dependencies just so we don't have trouble with stuff like this but I haven't looked into it deeply enough to be sure.

File: test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/ClassAnalyzer.java (hunk 8 above)
Comment (rmuir) [inline test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/ClassAnalyzer.java]: It tells you that. See the messages in Violation.java:
```
static enum Kind {        
        ESCAPES_WITHOUT_THROWING_ANYTHING("Control flow escapes with no exception"),
        THROWS_SOMETHING_ELSE_BUT_LOSES_ORIGINAL("Throws a different exception, losing the original"),
        ANNOTATED_AS_SWALLOWER_BUT_HAS_NO_PROBLEMS("Annotated with @SwallowsException but has no problems");
```

But this part is telling you its a wrong class annotation, versus a wrong method annotation. You can put the annotation in either place.

File: core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+ * under the License.
+ */
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation to suppress broken catch-block swallowing for a method or class
+ */
+@Target({ElementType.CONSTRUCTOR, ElementType.METHOD, ElementType.TYPE})
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java]: I think allowing this on a whole class is too broad. Is there a use case I'm not thinking of? I just figure it'd almost always be better to have it on a method or constructor.

File: test/tools/catch-analyzer/build.gradle [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+ * under the License.
+ */
+
+import org.elasticsearch.gradle.precommit.PrecommitTasks
+
+dependencies {
+  compile 'org.ow2.asm:asm-debug-all:6.0_ALPHA' // use asm-debug-all because WE LIKE LINE NUMBERS!
+  testCompile "com.carrotsearch.randomizedtesting:randomizedtesting-runner:${versions.randomizedrunner}"
+  testCompile "junit:junit:${versions.junit}"
+  testCompile "org.hamcrest:hamcrest-all:${versions.hamcrest}" // hamcrest is evil, but randomized runner no worky without
```
Comment (nik9000) [inline test/tools/catch-analyzer/build.gradle]: There are plenty of non evil uses of hamcrest. I think it'd be more accurate to say that this is a dependency of randomizedRunner but otherwise unused. Sadly, I don't believe there is a testRuntime scope.

File: core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java (hunk 10 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java]: Its the only way to annotate a clinit block.

**Comment by nik9000** [comment]
I reviewed the easy parts of the code (all but the asm bits) and I very much want it. I'm not sure what the right path is to get it merged though without failing the build for days while we fix all the issues it throws back. Maybe have a file where we store all of the known bad files and intentionally ignore them. That way we at least stop adding more of these. Then we work on removing them and finally remove support for (or just configuration for) the file? I know that is the strategy we used with line length and we still 1200 files that violate, but at least the list isn't growing any more and we're slowly removing them.

Working on this in a long running branch seems like a bad choice because you'll end up with a zillion conflicts as people continue to add things.

**Comment by rmuir** [comment]
Yes, that is my problem. I don't feel the scanner should be made more complex because the codebase is such a mess, but that is just me. Maybe we can add the thing as a standalone task (*not* called by precommit) as a step. Then it can be run manually and we can mark all the bad guys and keep up.

I've already had a branch for this thing for over a week and its very painful.

File: test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/ClassAnalyzer.java (hunk 8 above)
Comment (nik9000) [inline test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/ClassAnalyzer.java]: Got it.

File: core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java (hunk 10 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java]: Could we just force those blocks to call static methods if they really really want to swallow an exception?

**Comment by nik9000** [comment]
> I've already had a branch for this thing for over a week and its very painful.

Yeah. I can't imagine that working out well over a month or something.

A stand alone task doesn't stop us from adding more failures though, and that seems really important when migrating to a thing like that. It is better than a branch though. If you can get someone to review the analysis code I'm +1 on doing it as a stand alone task just to get it merged and then we can think about other solutions. I like the file idea because it is *fairly* fast to generate the files from failures but I understand about making it overcomplicated as a migration step.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java (hunk 2 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java]: Yeah. I was more thinking with my fingers.

File: core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java (hunk 10 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java]: you can. Its the same situation as forbidden apis.

Originally i developed this to use type annotations, e.g. `catch(@Swallowed Exception e) { ... `, but those have numerous problems:
* you can type-annotate an exception parameter in a clinit, it simply never makes it into the bytecode at all
* in the case of unreachable code, the annotation is also lost
* for finally blocks and try-with-resources statements, your code is duplicated, but not the annotation.

So basically, those are so unreliable, that we'd have to be lenient, and couldnt fail if the annotation was unnecessary. I think thats really important :)

**Comment by nik9000** [comment]
If you can write a script that adds the annotations from failures then we could add all the annotations with the script and merge it that way?

File: core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java (hunk 10 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java]: > fail if the annotation was unnecessary

Yeah, that is very important. It would be nice to be able to annotate at the exception level. Much cleaner if not for those nasty problems.

I still think we shouldn't allow the annotation on classes at all and should force them to make a static method call if they want to swallow. But I'm not so against it that I'd block this whole PR over it.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java]: That would require a fairly invasive refactoring of those clinit blocks. imagine code like this:

```
static final Foo x;
static final Bar y;

static {
  // sheisty shit that will initialize x and y
}
```

This is actually pretty common. Splitting up the large clinits into methods could arguably make the code more confusing in some of these cases, too.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java]: Yeah. I just don't like it! I'll live with it and maybe think of a way to make it less painful later.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/common/SwallowsExceptions.java]: I dont like it either. we should just remove the class part later IMO, after those things are refactored/cleaned up. It doesnt need to be allowed on classes FOREVER :)
